src/Build/Instance/ProjectInstance.cs

```diff
@@ -773,6 +786,18 @@ public string FullPath
             { return _itemDefinitions; }
         }
 
+        /// <summary>
+        /// The full file paths of all the files that during evaluation contributed to this project instance.
+        /// This does not include projects that were never imported because a condition on an Import element was false.
+        /// The outer ProjectRootElement that maps to this project instance itself is not included.
+        /// </summary>
+        public IReadOnlyList<string> Imports { get; private set; }
```

Comment:
Is there a reason these have to be read only? I don't know why you'd want to modify the list of imports, but I don't think it really hurts to give people the option, since that means you have to use `_imports.AsReadOnly()` instead of just `_imports`, for instance, which takes up a little extra time and space.

Reply:
It only makes sense as read only. If one we're to add to the collection, it wouldn't actually import anything for example.

Reply:
I understand, but it seems contradictory to say there's no reason to modify it (so it should be read only) but also modify it in RecordImport and RecordImportWithDuplicates.

If you have a way of importing more files after it's created, you might want to keep the list updated.

Reply:
It's read-only in the *public* interface. ie a consumer of MSBuild itself (the whole point of this change).

Reply:
`_imports` and `_importsIncludingDuplicates` are `List<string>`. Since the List type implements that interface there's no time/space concern calling `.AsReadOnly()`. And as David said giving external API consumers the a mutable `List` here would be the wrong choice.